source/utils/deep-freeze.ts

```diff
@@ -1,6 +1,6 @@
 import is from '@sindresorhus/is';
 
-export default function deepFreeze<T extends object>(object: T): Readonly<T> {
+export default function deepFreeze<T extends Record<string, any>>(object: T): Readonly<T> {
```

Comment:
This is also slightly stricter than `object`, which is why we should prefer it.

Reply:
```suggestion
export default function deepFreeze<T extends Record<string, unknown>>(object: T): Readonly<T> {
```

?

Reply:
It has to be `any` because it's spreading. It's a dumb TS thing. 

https://devblogs.microsoft.com/typescript/announcing-typescript-3-5/#higher-order-type-inference-from-generic-constructors See the `{ [k: string]: unknown } is no longer a wildcard assignment target` heading.

Reply:
I don't know what this exactly means, but this works:

```ts
import is from '@sindresorhus/is';

const isArray = is.array as (arg: unknown) => arg is unknown[];

export default function deepFreeze<T extends Record<string, unknown> | unknown[]>(object: T): Readonly<T> {
	for (const value of Object.values(object)) {
		if (is.plainObject(value) || isArray(value)) {
			deepFreeze(value);
		}
	}

	return Object.freeze(object);
}
```

and this does too:

```ts
import is from '@sindresorhus/is';

export default function deepFreeze<T extends Record<string, unknown>>(object: T): Readonly<T> {
	for (const value of Object.values(object)) {
		if (is.plainObject(value) || is.array(value)) {
			deepFreeze(value as Record<string, unknown>);
		}
	}

	return Object.freeze(object);
}
```

Reply:
1. Ah right. I missed that it also accepts arrays. Still fails though:

```
source/create.ts:292:59 - error TS2345: Argument of type 'InstanceDefaults' is not assignable to parameter of type 'unknown[] | Record<string, unknown>'.
  Type 'InstanceDefaults' is not assignable to type 'Record<string, unknown>'.
    Index signature is missing in type 'InstanceDefaults'.

292  	value: defaults.mutableDefaults ? defaults : deepFreeze(defaults),
```

2. Obviously that passes as you're using `as` which force-coerces it to whatever you want. That's not a proper solution though.

Reply:
Yeah, you're right. Let's go with any.